README.md
# TripleSpec Slit Viewer Focus Calculator

[![TripleSpec Focus](https://github.com/soar-telescope/triplespec_focus/actions/workflows/python-package.yml/badge.svg)](https://github.com/soar-telescope/triplespec_focus/actions/workflows/python-package.yml)
[![codecov](https://codecov.io/gh/soar-telescope/triplespec_focus/branch/main/graph/badge.svg?token=S6Y6R7AQNY)](https://codecov.io/gh/soar-telescope/triplespec_focus)
[![Upload to PYPI](https://github.com/soar-telescope/triplespec_focus/actions/workflows/python-publish.yml/badge.svg)](https://github.com/soar-telescope/triplespec_focus/actions/workflows/python-publish.yml)
[![Documentation Status](https://readthedocs.org/projects/triplespec-focus/badge/?version=latest)](https://triplespec-focus.readthedocs.io/en/latest/?badge=latest)
[![pypi](https://img.shields.io/pypi/v/triplespec_focus.svg?style=flat)](https://pypi.org/project/triplespec-focus/)
[![CodeQL](https://github.com/soar-telescope/triplespec_focus/actions/workflows/codeql-analysis.yml/badge.svg)](https://github.com/soar-telescope/triplespec_focus/actions/workflows/codeql-analysis.yml)
[![Python 3.8](https://img.shields.io/badge/Python-3.8-blue.svg)](https://www.python.org/downloads/release/python-380/)
[![GitHub license](https://img.shields.io/github/license/soar-telescope/triplespec_focus)](https://github.com/soar-telescope/triplespec_focus/blob/main/LICENSE)


![Test Coverage](https://codecov.io/gh/soar-telescope/triplespec_focus/branch/main/graphs/sunburst.svg?token=S6Y6R7AQNY)


TripleSpec is a fixed configuration Near Infrared Spectrograph. For more information please visit [NOIRLab Official Website](https://noirlab.edu/science/programs/ctio/instruments/triplespec41-nir-imaging-spectrograph) for
triplespec.

# Documentation

For documentation please visit the [ReadTheDocs Site](https://triplespec-focus.readthedocs.io/en/latest/?badge=latest) 

CHANGES.rst

.. _v1.0.0:

1.0.0 07-09-2022
================

- First functional version
- Uses ApertureStats to obtain sources' statistics
- Fully documented
- Fully tested

.. _v0.0.1.dev0:

0.0.1.dev0 06-05-2022
=====================

- Initial release of a non-functional tool but will all the packaging setup.


triplespec_focus/tests/test_utils.py
import os

import logging

from argparse import Namespace
from unittest import TestCase

import numpy as np
from astropy.modeling import models
from astropy.nddata import CCDData
from astropy.io import fits
from pandas import DataFrame

from ..utils import (circular_aperture_statistics,
                     get_args,
                     get_best_image_by_peak,
                     plot_sources_and_masked_data,
                     setup_logging)


class TestGetArgs(TestCase):

    def test_default_args_values(self):
        args = get_args(arguments=[])

        self.assertIsInstance(args, Namespace)
        self.assertEqual(args.data_path, os.getcwd())
        self.assertEqual(args.file_pattern, '*.fits')
        self.assertEqual(args.focus_key, 'TELFOCUS')
        self.assertEqual(args.filename_key, 'FILENAME')
        self.assertEqual(args.brightest, 5)
        self.assertEqual(args.saturation, 40000)
        self.assertEqual(args.source_fwhm, 7.0)
        self.assertEqual(args.detection_threshold, 6)
        self.assertEqual(args.mask_threshold, 1)
        self.assertFalse(args.plot_results)
        self.assertFalse(args.show_mask)
        self.assertFalse(args.debug)
        self.assertFalse(args.debug_plots)

    def test_modify_all_arguments(self):

        data_path = '/home/user/'
        file_pattern = '*.jpg'
        focus_key = 'FOCUS'
        filename_key = 'FILE'
        brightest = '10'
        saturation = '10000'
        source_fwhm = '40'
        detection_threshold = '3'
        mask_threshold = '3'

        new_args = ['--data-path', data_path,
                    '--file-pattern', file_pattern,
                    '--focus-key', focus_key,
                    '--filename-key', filename_key,
                    '--brightest', brightest,
                    '--saturation', saturation,
                    '--source-fwhm', source_fwhm,
                    '--detection-threshold', detection_threshold,
                    '--mask-threshold', mask_threshold,
                    '--plot-results',
                    '--show-mask',
                    '--debug',
                    '--debug-plots']

        args = get_args(arguments=new_args)
        self.assertIsInstance(args, Namespace)
        self.assertEqual(args.data_path, data_path)
        self.assertEqual(args.file_pattern, file_pattern)
        self.assertEqual(args.focus_key, focus_key)
        self.assertEqual(args.filename_key, filename_key)
        self.assertEqual(args.brightest, int(brightest))
        self.assertEqual(args.saturation, int(saturation))
        self.assertEqual(args.source_fwhm, source_fwhm)
        self.assertEqual(args.detection_threshold, detection_threshold)
        self.assertEqual(args.mask_threshold, mask_threshold)
        self.assertTrue(args.plot_results)
        self.assertTrue(args.show_mask)
        self.assertTrue(args.debug)
        self.assertTrue(args.debug_plots)


class TestLoggingSettings(TestCase):

    def test_debug_logging_setup(self):
        logger = setup_logging(debug=True)
        astropy_logger = logging.getLogger('astropy')
        self.assertEqual(logger.level, logging.DEBUG)
        self.assertTrue(astropy_logger.disabled)

    def test_enable_astropy_logger(self):
        logger = setup_logging(enable_astropy_logger=True)
        self.assertEqual(logger.level, logging.INFO)

    def test_default_logging_setup(self):
        logger = setup_logging()
        self.assertEqual(logger.level, logging.INFO)


class TestCircularApertureStatistics(TestCase):

    def setUp(self) -> None:
        self.positions = np.array(((100, 100), (300, 300)))
        self.aperture_radius = 10
        self.filename_key = 'FILE'
        self.focus_key = 'FOCUS'
        self.plot = False

        data_shape = (400, 400)
        noise = np.random.random_sample(size=data_shape)
        bias_level = 300.
        ones = np.ones(shape=data_shape)

        stars = None
        for x_mean, y_mean in self.positions:
            star_profile = models.Gaussian2D(amplitude=20000,
                                             x_mean=x_mean,
                                             y_mean=y_mean,
                                             x_stddev=self.aperture_radius / 2.,
                                             y_stddev=self.aperture_radius / 2.)
            if not stars:
                stars = star_profile
            else:
                stars += star_profile

        data = ones * bias_level + ones * stars.render(ones) + noise

        self.ccd = CCDData(data=data,
                           meta=fits.Header(),
                           unit='adu')
        self.ccd.header.set(self.filename_key,
                            value='some_file_name.fits',
                            comment='File name')
        self.ccd.header.set(self.focus_key,
                            value=1000,
                            comment='Focus value')

    def test_output(self):

        result = circular_aperture_statistics(
            ccd=self.ccd,
            positions=self.positions,
            aperture_radius=self.aperture_radius,
            filename_key=self.filename_key,
            focus_key=self.focus_key,
            plot=self.plot)

        self.assertIsInstance(result, DataFrame)
        self.assertTrue(all(column in result.columns.tolist() for column in ['id',
                                                                             'mean',
                                                                             'fwhm',
                                                                             'max',
                                                                             'xcentroid',
                                                                             'ycentroid',
                                                                             'focus',
                                                                             'filename']))


class TestFastBestImageDetection(TestCase):

    def setUp(self) -> None:
        peaks = [5000, 10000, 15000, 20000, 25000]
        focus = [500, 1000, 1500, 2000, 2500]
        self.saturation_level = 21000
        self.focus_key = 'FOCUS'

        self.file_list = []
        for i in range(1, 6):
            _file = f"test_file_{i}.fits"
            self.file_list.append(_file)
            data = np.ones(shape=(400, 400)) * peaks[i - 1]

            ccd = CCDData(data=data, header=fits.Header(), unit='adu')

            ccd.header.set('FILENAME', value=_file, comment="This file name")
            ccd.header.set('PEAK', value=peaks[i - 1], comment="This file set peak value")
            ccd.header.set(self.focus_key, value=focus[i - 1], comment="This file set peak value")
            ccd.write(_file, overwrite=True)


    def tearDown(self) -> None:
        for _file in self.file_list:
            if os.path.isfile(_file):
                os.unlink(_file)

    def test_get_best_image_by_peak(self):
        best_image = get_best_image_by_peak(file_list=self.file_list,
                                            saturation_level=self.saturation_level,
                                            focus_key=self.focus_key)
        image_name, peak, focus = best_image

        self.assertIsInstance(best_image, list)
        self.assertEqual(len(best_image), 3)
        self.assertEqual(image_name, 'test_file_4.fits')
        self.assertEqual(peak, 20000.)
        self.assertEqual(focus, 2000)




triplespec_focus/tests/test_triplespec_focus.py
import os
import requests

from astropy.modeling import Model
from astropy.table import QTable
from ccdproc import CCDData
from pandas import DataFrame
from pathlib import Path
from unittest import TestCase

from ..triplespec_focus import TripleSpecFocus


class TripleSpecFocusTest(TestCase):

    def setUp(self) -> None:
        self.n_brightest_sources = 10
        self.tspec_focus = TripleSpecFocus()

        self.zenodo_id = '7055458'

        self.all_files = [
            'SV_ARC_13-04-2022_0010.fits',
            'SV_ARC_13-04-2022_0012.fits',
            'SV_ARC_13-04-2022_0014.fits',
            'SV_ARC_13-04-2022_0016.fits',
            'SV_ARC_13-04-2022_0019.fits',
            'SV_ARC_13-04-2022_0021.fits',
            'SV_ARC_13-04-2022_0023.fits',
            'SV_ARC_13-04-2022_0025.fits',
            'SV_ARC_13-04-2022_0011.fits',
            'SV_ARC_13-04-2022_0013.fits',
            'SV_ARC_13-04-2022_0015.fits',
            'SV_ARC_13-04-2022_0018.fits',
            'SV_ARC_13-04-2022_0020.fits',
            'SV_ARC_13-04-2022_0022.fits',
            'SV_ARC_13-04-2022_0024.fits',
            'SV_ARC_13-04-2022_0026.fits']

        self.test_file_name = 'SV_ARC_13-04-2022_0013.fits'

        for _file in self.all_files:
            if not os.path.exists(_file):
                url = f"https://zenodo.org/record/{self.zenodo_id}/files/{_file}?download=1"
                response = requests.get(url)
                with open(_file, "wb") as test_file:
                    print(f"Downloading file: {_file}")
                    test_file.write(response.content)

    def test_instance_is_correct(self):
        self.assertIsInstance(self.tspec_focus, TripleSpecFocus)

    def test_call_method_whithout_data_path_nor_file_list(self):
        self.assertRaises(SystemExit, self.tspec_focus)

    def test_get_focus_with_no_data(self):
        df = DataFrame()
        self.assertRaises(KeyError, self.tspec_focus.get_best_focus, df)

    def test_get_focus_with_data(self):
        focus_values = [-1300, -1200, -1100, -1000, -900, -800, -700]
        fwhm_values = [9, 8, 7, 6, 7, 8, 9]
        data = {'focus': focus_values, 'fwhm': fwhm_values}
        df = DataFrame(data=data)
        self.assertIsNone(self.tspec_focus.best_focus)
        self.assertIsNone(self.tspec_focus.fitted_model)
        results = self.tspec_focus.get_best_focus(df=df, x_axis_size=2000)
        self.assertAlmostEqual(self.tspec_focus.best_focus, -1000, places=4)
        self.assertIsNotNone(self.tspec_focus.fitted_model)
        self.assertIsInstance(self.tspec_focus.fitted_model, Model)
        self.assertIsInstance(results, list)

    def test_detect_sources(self):
        ccd = CCDData.read(self.test_file_name, unit='adu')
        self.tspec_focus.show_mask = True
        sources = self.tspec_focus.detect_sources(ccd=ccd, debug_plots=False)
        self.assertIsInstance(sources, QTable)
        self.assertEqual(len(sources), 94)

    def test_call_with_data_path(self):

        results = self.tspec_focus(data_path='./')
        self.assertIsInstance(results, dict)
        self.assertEqual(len(results), 11)

    def test_call_with_file_list(self):
        data_path = Path('./')

        results = self.tspec_focus(file_list=sorted(data_path.glob(pattern='*.fits')))
        self.assertIsInstance(results, dict)
        self.assertEqual(len(results), 11)




triplespec_focus/entrypoints.py
import json

from .utils import get_args
from .triplespec_focus import TripleSpecFocus


def run_triplespec_focus(args=None):

    args = get_args(arguments=args)

    focus = TripleSpecFocus(debug=args.debug,
                            focus_key=args.focus_key,
                            filename_key=args.filename_key,
                            file_pattern=args.file_pattern,
                            n_brightest=args.brightest,
                            saturation=args.saturation,
                            plot_results=args.plot_results,
                            debug_plots=args.debug_plots)

    results = focus(data_path=args.data_path,
                    source_fwhm=args.source_fwhm,
                    det_threshold=args.detection_threshold,
                    mask_threshold=args.mask_threshold,
                    n_brightest=args.brightest,
                    saturation_level=args.saturation,
                    show_mask=args.show_mask,
                    plot_results=args.plot_results,
                    debug_plots=args.debug_plots,
                    print_all_data=False)
    print(json.dumps(results, indent=4))


if __name__ == '__main__':
    run_triplespec_focus()


triplespec_focus/utils.py
import copy
import os
import logging

import numpy as np
import numpy.ma as ma

from argparse import ArgumentParser, Namespace

from astropy.visualization import ZScaleInterval
from ccdproc import CCDData
from logging import Logger

from matplotlib import cm, pyplot as plt
from pandas import DataFrame
from typing import Union, List

from photutils import CircularAperture, CircularAnnulus, ApertureStats

log = logging.getLogger(__name__)


def circular_aperture_statistics(ccd: CCDData,
                                 positions: np.ndarray,
                                 aperture_radius: float = 10,
                                 filename_key: str = 'FILENAME',
                                 focus_key: str = 'TELFOCUS',
                                 plot: bool = False) -> DataFrame:
    """Obtain aperture statistics from a FITS file with point sources

    Uses CircularAperture from photutils to obtain several data but the most important is the FWHM

    Args:
        ccd (CCDData): An image with point sources.
        positions (numpy.ndarray): Coordinates of the sources to be measured.
        aperture_radius (float): Aperture size for obtaining measurements. Default 10.
        filename_key (str): FITS keyword name for obtaining the file name from the FITS file. Default FILENAME.
        focus_key (str): FITS keyword name for obtaining the focus value from the FITS file. Default TELFOCUS.
        plot (bool): If set to True will display a plot of the image and the measured sources.

    Returns:
        A DataFrame containing the following columns: id, mean, fwhm, max, xcentroid, ycentroid and filename.

    """
    apertures = CircularAperture(positions, r=aperture_radius)

    aper_stats = ApertureStats(ccd.data, apertures, sigma_clip=None)

    columns = ('id', 'mean', 'fwhm', 'max', 'xcentroid', 'ycentroid')
    as_table = aper_stats.to_table(columns=columns)
    as_table['focus'] = ccd.header[focus_key]
    as_table['filename'] = ccd.header[filename_key]
    for col in as_table.colnames:
        if col not in ['filename']:
            as_table[col].info.format = '%.8g'

    if plot:  # pragma: no cover
        title = f"Photometry of {ccd.header['FILENAME']}"
        plot_sources_and_masked_data(ccd=ccd, positions=positions, title=title)

    return as_table.to_pandas()


def get_args(arguments: Union[List, None] = None) -> Namespace:
    """Helper function to get the console arguments using argparse.

    Args:
        arguments (List, None): Optional list of arguments. Default None.

    Returns:
        An instance of arparse's Namespace.

    """
    parser = ArgumentParser(
        description="Get best focus value using a sequence of images with "
                    "different focus value"
    )

    parser.add_argument('--data-path',
                        action='store',
                        dest='data_path',
                        default=os.getcwd(),
                        help='Folder where data is located')

    parser.add_argument('--file-pattern',
                        action='store',
                        dest='file_pattern',
                        default='*.fits',
                        help='Pattern for filtering files.')

    parser.add_argument('--focus-key',
                        action='store',
                        dest='focus_key',
                        default='TELFOCUS',
                        help='FITS header keyword to find the focus value.')

    parser.add_argument('--filename-key',
                        action='store',
                        dest='filename_key',
                        default='FILENAME',
                        help='FITS header keyword to find the current file name.')

    parser.add_argument('--brightest',
                        action='store',
                        dest='brightest',
                        type=int,
                        default=5,
                        help='Pick N-brightest sources to perform statistics. Default 5.')

    parser.add_argument('--saturation',
                        action='store',
                        dest='saturation',
                        type=int,
                        default=40000,
                        help='Saturation value for data')

    parser.add_argument('--source-fwhm',
                        action='store',
                        dest='source_fwhm',
                        default=7.0,
                        help='FWHM for source detection.')

    parser.add_argument('--detection-threshold',
                        action='store',
                        dest='detection_threshold',
                        default=6,
                        help='Number of standard deviation above median for source detection.')

    parser.add_argument('--mask-threshold',
                        action='store',
                        dest='mask_threshold',
                        default=1,
                        help='Number of standard deviation below median to mask values.')

    parser.add_argument('--plot-results',
                        action='store_true',
                        dest='plot_results',
                        help='Show a plot when it finishes the focus '
                             'calculation')

    parser.add_argument('--show-mask',
                        action='store_true',
                        dest='show_mask',
                        help='Show the image and the masked areas highlighted in red.')

    parser.add_argument('--debug',
                        action='store_true',
                        dest='debug',
                        help='Activate debug mode')

    parser.add_argument('--debug-plots',
                        action='store_true',
                        dest='debug_plots',
                        help='Show debugging plots.')

    args = parser.parse_args(args=arguments)
    return args


def get_best_image_by_peak(file_list: List, saturation_level: float = 40000., focus_key: str = 'TELFOCUS') -> List:
    """Select the best image by its peak value

    The best peak must be the highest below the saturation level, therefore the data is first masked and then sorted
    by the peak.

    Args:
        file_list (List): The list of images to be used.
        saturation_level (float): Data level at which the image saturates. Default 40000.
        focus_key (str): FITS keyword name for obtaining the focus value from the FITS file. Default TELFOCUS.

    Returns:
        A list wit the best image according to the criteria described above.

    """
    data = []
    for f in file_list:
        ccd = CCDData.read(f, unit='adu')
        ccd.mask = ccd.data >= saturation_level
        masked_data = ma.masked_array(ccd.data, mask=ccd.mask)
        np_max = ma.MaskedArray.max(masked_data)
        if not ma.is_masked(np_max):
            log.debug(f"File: {os.path.basename(f)} Max: {np_max} Focus: {ccd.header[focus_key]}")
            data.append([f, np_max, ccd.header[focus_key]])
        else:
            log.debug(f"Rejected masked value {np_max} from file {f}")

    df = DataFrame(data, columns=['file', 'peak', 'focus'])
    best_image = df.iloc[df['peak'].idxmax()]
    log.info(f"Best Image: {best_image.file} Peak: {best_image.peak} Focus: {best_image.focus}")

    return best_image.to_list()


def plot_sources_and_masked_data(ccd: CCDData, positions: np.ndarray, title: str = '', aperture_radius: int = 10):  # pragma: no cover
    """Helper function to plot data and sources

    Args:
        ccd (CCDData): The image to be shown.
        positions (numpy.ndarray): Array of source positions to be plotted over the image.
        title (str): Title to put to the plot. Default '' (empty string).
        aperture_radius (int): Radius size in pixels for the drawing of the sources. Default 10.

    """
    fig, ax = plt.subplots(figsize=(16, 9))
    ax.set_title(title)

    color_map = copy.copy(cm.gray)
    color_map.set_bad(color='green')

    scale = ZScaleInterval()
    z1, z2 = scale.get_limits(ccd.data)

    apertures = CircularAperture(positions, r=aperture_radius)
    annulus_apertures = CircularAnnulus(positions, r_in=aperture_radius + 5, r_out=aperture_radius + 10)

    ax.imshow(ccd.data, cmap=color_map, clim=(z1, z2))
    apertures.plot(color='red')
    annulus_apertures.plot(color='yellow')

    plt.tight_layout()
    plt.show()


def setup_logging(debug: bool = False, enable_astropy_logger: bool = False) -> Logger:
    """Helper function to setup the logger.

    Args:
        debug (bool): If set to True will create a logger in debug level. Default False.
        enable_astropy_logger (bool): If set to True will allow the astropy logger to remain active. Default False.

    Returns:
        A Logger instance from python's logging.

    """
    if debug:
        log_level = logging.DEBUG
    else:
        log_level = logging.INFO

    logger = logging.getLogger(__name__)
    logger.setLevel(level=log_level)

    console = logging.StreamHandler()
    console.setLevel(log_level)

    log_format = '[%(asctime)s][%(levelname)s]: %(message)s'
    formatter = logging.Formatter(log_format)

    console.setFormatter(formatter)

    logger.addHandler(console)

    astropy_logger = logging.getLogger('astropy')
    if enable_astropy_logger or debug:
        for handler in astropy_logger.handlers:
            astropy_logger.removeHandler(handler)
        astropy_logger.addHandler(console)
    else:
        astropy_logger.disabled = True

    return logger
